omegaconf/nodes.py

```diff
@@ -40,16 +40,22 @@ def _set_value(self, value: Any, flags: Optional[Dict[str, bool]] = None) -> Non
         ):
             self._val = value
         else:
-            if not self._metadata.optional and value is None:
-                raise ValidationError("Non optional field cannot be assigned None")
             self._val = self.validate_and_convert(value)
 
     def validate_and_convert(self, value: Any) -> Any:
         """
         Validates input and converts to canonical form
         :param value: input value
-        :return:  converted value ("100" may be converted to 100 for example)
+        :return: converted value ("100" may be converted to 100 for example)
         """
+        if value is None:
+            if self._is_optional():
+                return None
+            raise ValidationError("Non optional field cannot be assigned None")
+        # Subclasses can assume that `value` is not None in `_validate_and_convert_impl()`.
+        return self._validate_and_convert_impl(value)
```

Comment:
What do you think about adding `assert value is not None` to each of the `_validate_and_convert_impl` methods?
My motivation for this idea is to help catch future bugs as the OmegaConf internals evolve.

Reply:
I don't mind either way -- let's have @omry tell us what he likes best :)

Reply:
I don't think it's too useful because _validate_and_convert_impl is only called from that one callsite and I am not expecting it to be called directly.